README.md
# ou_Axion_limit
ou_Axion_limit is a package for calculate the Axion limit and analyse normal distribution

# Download
`
pip install ou-Axion-limit
`
# Usage

## Glimit

Input you experiment parameter as this way
```
from ou_Axion_limit import Glimit
g = Glimit()
g.B    = 8
g.Q    = 30000
g.T    = 2
g.total_time = 14 * 24 * 3600
g.SNR  = 5
g.f    = 4.74e9
g.beta = 2
g.cooling = 5 * 60
g.delta_w = 5000
g.delta_v = 1000
```
### g.information()
It gives you the limit and few parameters
```
g.information()

| ================== Parameter ==============
| f =  4.740e+09 (Frequency [Hz])
| B =      8.000 (Magnetic[T])
| V =  2.341e-04 (Cavity Volume [L])
| Q =  30000.000 (Q factor)
| T =      2.000 (Noise temp [k])
| t = 1209300.000 (Integration time [s])
| ================== OUR (SNR = 5) ==========
| g_a_gamma = 2.1790437845360718e-14 GeV^-1
| g_gamma   = 2.9111980151898598
| ================== KSVZ (SNR = 5) =========
| ksvz_g_a_gamma = 7.26049021733117e-15 GeV^-1
| ksvz_g_gamma   = -0.97
| ===========================================
```
### g.to_g_gamma(x)
convert G_a_gamma_gamma to G_gamma
```
print(g.to_g_gamma(1.3e-13))
7.367
```
### g.find_limit(10)
Find the integration time with given numbers of KSVZ limit (10 is means 10 times G_KSVZ ), and it will plot a figure to you
```
g.find_limit(10)

| ============== Parameter ==================
| f =  4.740e+09 (Frequency [Hz])
| B =      8.000 (Magnetic[T])
| V =  2.341e-04 (Cavity Volume [L])
| Q =  30000.000 (Q factor)
| T =      2.000 (Noise temp [k])
| ===========================================


[*] Total time :  1209600
[*] Seaching limit : 10 times KSVZ limit
[*] Founded Answer :
	 Integration :   20.21 [minutes] to reach 10 times KSVZ limit
	 Cold down   :  300.00 [seconds]
	 Total time  :  336.00 [hours]
	 Step size   :   79.00 [kHz]
	 Move rod    :  799.53 [steps] (should be a convert to a integer)
	 Total scan  :   63.16 [MHz]
```

## analyse

import some useful package

```
from matplotlib.pyplot import *
from numpy import *
from ou_Axion_limit import analyse
```

analyse a normal distribution

### analyse(data)

```
data = np.random.normal(1,1,2000)
an = analyse(data)
print(an.mu, an.sigma)
```
### an.histogram(cut)

Plot the histogram with matplotlib

```
figure()
subplot(211)
title("data")
plot(data)
subplot(212)
an.histogram(cut = 1.645)
show()
```

CHANGELOG.txt
Change log
===========

1.0.0 (06/10/2021)
-------------------
- fix :
	fix some bugs


ou_Axion_limit/Analy.py
from matplotlib.pyplot import *
from numpy import *
from scipy.stats import norm, chisquare, poisson, chi2
from scipy.optimize import curve_fit


class analyse:
    def __init__(self,data,dbm = False):
        self.data = data
        if (dbm):
            self.data  = 10**(self.data/10)
        self.norm_fittting()
        
    def norm_fittting(self):


        n, bins    = histogram(self.data,100,density=True)
        centers    = 0.5*(bins[1:]+bins[:-1])
        P0         = [mean(self.data), sqrt(mean((self.data-mean(self.data))**2))]
        try:
            pars, cov           = curve_fit(lambda x, mu, sig : norm.pdf(x, loc=mu, scale=sig), centers, n, P0)
            self.mu, self.sigma = pars
            self.mu_error       = sqrt(cov[0,0])
            self.sigma_error    = sqrt(cov[1,1])
        except:
            self.mu, self.sigma = P0
            self.mu_error       = 0
            self.sigma_error    = 0

        # (self.mu, self.sigma) = norm.fit(self.data)

    def chisqure(self,x,y):
        return mean((x-y)**2/y)

    def histogram(self,cut = None):
        n, bins, patches  = hist(self.data,100,density=True)
        xx_middle = 0.5 * (bins[1:] + bins[:-1])
        if (cut):
            this_cut = self.mu + cut
        y2 = norm.pdf(xx_middle, self.mu, self.sigma)

        l = plot(xx_middle, y2, 'r--', linewidth=2)
        title(r'$\mathrm{Histogram\ of\ data:}\ \mu=%.4f\pm%.4f,\ \sigma=%.4f\pm%.4f$' %(
            self.mu,
            self.mu_error,
            self.sigma,
            self.sigma_error))
        print(f"chisqure = {chisquare(n,y2)}")
        if (cut):
            plot([this_cut,this_cut], [min(n),max(n)],label = f"{cut}"+r"$\sigma$")
        tight_layout()
        show()
        return n, xx_middle, patches


if __name__ == "__main__":
    data = np.random.normal(1,1,2000)
    an = analyse(data)
    print(an.mu, an.sigma)
    print(mean(data),sqrt(mean((data-mean(data))**2)))
    an.histogram(1.645)

ou_Axion_limit/Glimit.py
from matplotlib.pyplot import *
from numpy import *



class Glimit:
    def __init__(self):
        self.h      = 6.626e-34 # J/k
        self.h_bar  = 4.135e-15/(2*pi) # ev
        self.k      = 1.38e-23 # J
        self.rho    = 0.45e15   # axion density (GeV/cc)
        self.big_A  = 78e6  #   78Mev
        self.C      = 3e8    # The Speed of light
        self.alpha  = 1/137  # fine structure const
        self.mu     = 4*pi * 1e-7 
        self.e      = 1.6e-19
        self.he     = self.h/self.e
        self.cm     = 0.6527    # one hour (s)

        self.s          = 0      # S₁₁
        self.T          = .3     # physical + noise (k) 
        self.f          = 5e9    # 5Ghz Frequency (10⁹ Hz)
        self.B          = 8      # Magnetic (T)
        self.V          = 120*pi*621e-9 #48338e-9 # Volume (L)
        self.Q          = 25000 # Q factor 
        self.delta_v    = 1000   # 5kHz
        self.delta_w    = 5000   # 5kHz
        self.cooling    = 0
        self.SNR        = 5
        self.beta       = 1
        self.total_time = 86400 * 7
        self.Scanwin    = 2e6
        self.Scanran    = 2e6
        self.Scanshi    = 2e6
        
        self.ksvz_g_gamma = -0.97
        self.dfsz_g_gamma = 0.36
        self.calculate()

    def calculate(self):
        
        self.w         = 2*pi*(self.f)
        self.ma        = self.h_bar * self.w
        self.t         = self.total_time / self.Scanran * self.Scanshi - self.cooling

        assert self.t  > 0,"No time for cooling down"

        self.g_KSVZ    = 0.97 * self.ma * self.alpha /(pi * self.big_A * self.big_A)
        self.Na        = self.delta_v * self.t
        self.Ns        = self.Scanwin/self.Scanshi
        self.sigma     = self.k * self.T * self.delta_w/sqrt(self.Ns * self.Na) 
        
        self.shift =  ((self.h_bar*self.C)**3)*self.rho / (self.ma**2)  * (1/self.mu) * (self.B**2) * \
        self.V * self.w* self.cm * self.Q *self.beta / (1+self.beta)
        
    def cal_single_point(self):
        self.w     = 2*pi*(self.f)
        self.ma    = self.h_bar * self.w
        self.Na    = self.delta_v * self.t
        self.Ns    = 1
        self.sigma = self.k * self.T * self.delta_w/sqrt(self.Na) 
        self.shift =  ((self.h_bar*self.C)**3)*self.rho / (self.ma**2)  * (1/self.mu) * (self.B**2) * \
        self.V * self.w* self.cm * self.Q *self.beta / (1+self.beta)

    def g_a_gamma(self):
        return sqrt( self.SNR*self.sigma/self.shift) * 1e9
    
    def g_gamma(self):
        return  (pi * self.big_A * self.big_A) * self.g_a_gamma() / self.ma / self.alpha   * 1e-9    
    
    # convert G_a_gamma_gamma to G_gamma
    def to_g_gamma(self,x):
        return (pi * self.big_A * self.big_A) * x / self.ma / self.alpha   * 1e-9


    def ksvz_g_a_gamma(self):
        return 0.97 * self.ma * self.alpha /(pi * self.big_A * self.big_A)  * 1e9
        
    def information(self):
        self.calculate()
        print("|","="*18,"Parameter","="*14)
        print(f"| f = {self.f:10.3e} (Frequency [Hz])")
        print(f"| B = {self.B:10.3f} (Magnetic[T])")
        print(f"| V = {self.V:10.3e} (Cavity Volume [L])")
        print(f"| Q = {self.Q:10.3f} (Q factor)")
        print(f"| T = {self.T:10.3f} (Noise temp [k])")
        print(f"| t = {self.t:10.3f} (Integration time [s])")
        print("|","="*18,f"OUR (SNR = {self.SNR:d})","="*10)
        print(f"| g_a_gamma = {self.g_a_gamma() } GeV^-1")
        print(f"| g_gamma   = {self.g_gamma()}")
        print("|","="*18,f"KSVZ (SNR = {self.SNR:d})","="*9)
        print("|",f"ksvz_g_a_gamma = {self.ksvz_g_a_gamma()} GeV^-1")
        print("|",f"ksvz_g_gamma   = {self.ksvz_g_gamma}")
        print("|","="*43)
        print()

    # Find the  scan range and integration time with given target limit
    def find_limit(self,target):
        
        print("|","="*14,"Parameter","="*18)
        print(f"| f = {self.f:10.3e} (Frequency [Hz])")
        print(f"| B = {self.B:10.3f} (Magnetic[T])")
        print(f"| V = {self.V:10.3e} (Cavity Volume [L])")
        print(f"| Q = {self.Q:10.3f} (Q factor)")
        print(f"| T = {self.T:10.3f} (Noise temp [k])")
        print("|","="*43)
        print("\n")
        print("[*] Total time : ",self.total_time)
        print("[*] Seaching limit :",target ,"times KSVZ limit")

        def start_exp(scan_shfit):
            spectrum    = zeros(int((scan_range + scan_windos)//grid_unit))
            weight      = zeros(int((scan_range + scan_windos)//grid_unit))
            sigma       = zeros(int((scan_range + scan_windos)//grid_unit))
            shift_index = int(scan_shfit/grid_unit)
            start_index = 0
            
            f_now = f_start
            noise        = random.normal(0,1,windo_index)
            while f_now <= f_end:
                noise        = np.zeros(windo_index) +1
                scan_f       = linspace(f_now - scan_windos/2, f_now + scan_windos/2, windo_index)
                sigma_this   = 1
                
                w   = Lorentz(scan_f, f_now, self.Q) / sigma_this**2
                noise *= w
                sig = sigma_this**2 * (w)**2
                
                spectrum[start_index:start_index+windo_index] = spectrum[start_index:start_index+windo_index] + noise*w
                weight[  start_index:start_index+windo_index]   = weight[start_index:start_index+windo_index] + w
                sigma[   start_index:start_index+windo_index]    = sigma[start_index:start_index+windo_index] + sig
                
                f_now       += scan_shfit
                start_index += shift_index
            return linspace(f_start, f_end, int((scan_range + scan_windos)//grid_unit)), spectrum, sigma

        def Lorentz(x,fr,Q):
            return 1 / (1 + ( 2*Q*(x/fr-1))**2)
        def merging(x,y,merge):
            out_x, out_y = [], []
            for i in range(len(x)//merge):
                out_x.append(mean(x[i*merge:(i+1)*merge]))
                out_y.append(sum( y[i*merge:(i+1)*merge]))
            return array(out_x), array(out_y)

        grid_unit   = 1e3
        f_start     = self.f - 200e3*100
        f_end       = self.f + 200e3*100
        scan_windos = 2e6
        windo_index = int(scan_windos/grid_unit)
        scan_range  = f_end - f_start
        start_index = int((scan_windos/2) / grid_unit)
        end_index   = start_index + int(scan_range / grid_unit)

        x,spec, sigm = start_exp(self.f / self.Q / 2)
        # print(spec,sigm)
        
        _, spec = merging(x,spec,5)
        x, sigm = merging(x,sigm,5)
        y = np.divide(spec, sqrt(sigm), out=np.zeros_like(spec), where=sqrt(sigm)!=0)

        self.t = 600*60
        self.cal_single_point()
        with np.errstate(divide='ignore'):
            g_a_gamma = sqrt(self.sigma*5/self.shift/y) * 1e9

        g_gamma = self.to_g_gamma(g_a_gamma)

        #  (pi * self.big_A * self.big_A) * g_a_gamma / self.ma / self.alpha   * 1e-9
        # print(min(g_gamma/0.97))

        new_t = self.t * ( min(g_gamma/0.97) / target)  **4
        print(f"[*] Founded Answer :")
        print(f"\t Integration : {(new_t/60):7.2f} [minutes] to reach {target} times KSVZ limit")
        print(f"\t Cold down   : {self.cooling:7.2f} [seconds]")
        print(f"\t Total time  : {self.total_time/3600:7.2f} [hours]")
        print(f"\t Step size   : {self.f / self.Q / 2*1e-3:7.2f} [kHz]")
        print(f"\t Move rod    : {self.total_time / (new_t + self.cooling):7.2f} [steps] (should be a convert to a integer)")
        print(f"\t Total scan  : {self.total_time / (new_t + self.cooling) * self.f / self.Q / 2*1e-6:7.2f} [MHz]")


        self.t = new_t

        assert new_t < self.total_time ,"Integration time is larger then total time"


        f_start     = self.f - self.f / self.Q / 2*(self.total_time / (new_t + self.cooling)//2)
        f_end       = self.f + self.f / self.Q / 2*(self.total_time / (new_t + self.cooling)//2)
        scan_range  = f_end - f_start
        start_index = int((scan_windos/2) / grid_unit)
        end_index   = start_index + int(scan_range / grid_unit)
        x,spec, sigm = start_exp(self.f / self.Q / 2)
        _, spec = merging(x,spec,5)
        x, sigm = merging(x,sigm,5)
        y = np.divide(spec, sqrt(sigm), out=np.zeros_like(spec), where=sqrt(sigm)!=0)

        self.cal_single_point()
        with np.errstate(divide='ignore'):
            g_a_gamma = sqrt(self.sigma*5/self.shift/y) * 1e9
        g_gamma = (pi * self.big_A * self.big_A) * g_a_gamma / self.ma / self.alpha   * 1e-9

        figure()
        max_ratio = min(g_gamma/0.97)
        title(f"Best value : {max_ratio:.3f} ")
        plot(x*1e-9,g_gamma/0.97,label="This experiment")

        plot(x*1e-9,abs(self.ksvz_g_gamma +x*0)/0.97,"b--",label="KSVZ")
        plot(x*1e-9,abs(self.dfsz_g_gamma +x*0)/0.97,"r--",label="DFSZ")
        upper = 4
        down  = abs(self.dfsz_g_gamma +x*0)/abs(self.ksvz_g_gamma) / 4
        fill_between(x*1e-9,upper,down,color="yellow",label="model region")

        xlabel(f"Freq [GHz]")
        ylabel(r"$\frac{G_\gamma}{G_{KSVZ}}$",size=20)
        xlim(min(x*1e-9),max(x*1e-9))
        tight_layout()
        grid()
        gcf().autofmt_xdate()
        legend()
        tight_layout()
        show()


if __name__ == "__main__":
    g = Glimit()
    g.B    = 8
    g.Q    = 30000
    g.T    = 2
    g.total_time = 14 * 24 * 3600
    g.SNR  = 5
    g.f    = 4.74e9
    g.beta = 2
    g.cooling = 5 * 60
    g.delta_w = 5000
    g.delta_v = 1000
    g.information()
    # print(g.to_g_gamma(1.3e-13))

    g.find_limit(10)